server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java

```diff
@@ -85,18 +145,70 @@ public String[] excludes() {
         return this.excludes;
     }
 
+    public Boolean includeVectors() {
+        return this.includeVectors;
+    }
+
+    public boolean filterVectorFields() {
+        return this.includeVectors == null || this.includeVectors == false;
+    }
+
     public boolean hasFilter() {
         return this.includes.length > 0 || this.excludes.length > 0;
     }
 
     public SourceFilter filter() {
-        return new SourceFilter(includes, excludes);
+        return filter(null);
+    }
+
+    public SourceFilter filter(MappingLookup mappingLookup) {
+        if (filterVectorFields() == false) {
+            return new SourceFilter(includes, excludes);
+        } else {
+            if (mappingLookup == null) {
+                throw new IllegalStateException("mappingLookup must not be null when filtering vectors");
+            }
+
+            String[] excludeFields = excludes();
+            List<String> inferenceFieldNames = new ArrayList<>(mappingLookup.inferenceFields().size());
+            for (String inferenceField : mappingLookup.inferenceFields().keySet()) {
+                Mapper mapper = mappingLookup.getMapper(inferenceField);
+                if (mapper instanceof InferenceFieldMapper == false) {
+                    throw new IllegalStateException("Mapper for field [" + inferenceField + "] is not an inference field mapper");
+                }
+
+                InferenceFieldMapper inferenceFieldMapper = (InferenceFieldMapper) mapper;
+                inferenceFieldNames.add(inferenceFieldMapper.getInferenceFieldName());
```

Comment:
The original intent for `include_vectors` was to exclude only the embedding fields, not the entire semantic text field. The parameter should be limited to vector fields so that features like inner_hits can still access the semantic text chunks, for instance.

Could we add a flag to the dense/sparse vector field, such as `DenseVectorFieldMapper#includeSourceDefault`? Then, the semantic text field could set this flag to false when creating the embedding field.

Reply:
It's not only about inner hits. This feature is called `include_vectors`, so I'd prefer to limit its scope to dense and sparse vectors only. It's fine to set a different default for vector fields created by semantic text fields. Chunks are valuable information with **and** without vectors.

Reply:
To clarify, whether we filter out the entire `inference` field or just the embeddings has no impact on what we can return with `inner_hits`. We can (and do) build an entirely separate `FetchSourceContext` object for inner hits retrieval that currently filters out only the embeddings so that only the chunk text is returned.

Reply:
Thanks for clarifying. My main point is not on the `inner_hits` option though. It's about being consistent. We agreed that the layout of the semantic text field is not an implementation detail, it's the result of a transformation.
`include_vectors` is about returning vectors or not, it should have nothing to do with the hiding of the semantic text object imo.

Reply:
If we're filtering only the `embeddings` field, that also means the `model_settings` field will remain in `_source`. Do we want to filter that out as well?

Reply:
That's my main point, I'd prefer to keep the option only for vector fields. Why would this option remove the `model_settings`? It's unrelated and not a problem for users since the section is small. The main issue with vector fields is that they are large so they should not be returned by default. Hiding the layout of the semantic text field is a different issue that I wish we don't try to solve here.